### DISCUSSION TITLE
parsing: switching from json to ordered_json

### DISCUSSION THREAD

**Discussion by robinsch** [nlohmann/json#3731] [Q&A] [answered]
Imagine I have this code piece which parses a json to a lua table:

```cpp
void settable_from_json(LuaState* state, nlohmann::json json)
{
	lua_createtable(state, 0, 0);
	lua_checkstack(state, 2);

	for (auto const& item : json.items())
		parse_json_item(item, state);
}
```

parse_json_item contains this and can call itself recursivly to parse arrays/objects:
```
void parse_json_item(nlohmann::json_abi_v3_11_2::detail::iteration_proxy_value<nlohmann::json_abi_v3_11_2::detail::iter_impl<nlohmann::json_abi_v3_11_2::json>> const& val, LuaState* state)
{
	// key
	if (is_number(val.key()))
		lua_pushinteger(state, std::atoi(val.key().c_str()) + 1);
	else
		lua_pushstring(state, val.key().c_str());

	// value
	switch (val.value().type())
	{
		case nlohmann::json::value_t::array:
		case nlohmann::json::value_t::object:
		{
			lua_createtable(state, 0, 0);
			lua_checkstack(state, 2);

			for (auto const& val : val.value().items())
				parse_json_item(val, state);
			break;
		}
		case nlohmann::json::value_t::boolean:
			lua_pushboolean(state, val.value().get<bool>());
			break;
		case nlohmann::json::value_t::string:
			lua_pushstring(state, val.value().get<std::string>().c_str());
			break;
		case nlohmann::json::value_t::number_integer:
		case nlohmann::json::value_t::number_unsigned:
			lua_pushinteger(state, val.value().get<int32>());
			break;
		case nlohmann::json::value_t::number_float:
			lua_pushnumber(state, val.value().get<float>());
			break;
		default:
			lua_pushnil(state);
	}

	lua_settable(state, -3);
}
```

When changing the type to ```ordered_json``` in ```settable_from_json``` and ```parse_json_item``` I get greeted by this error on MSVC 19:

```
Error	C2039	'difference_type': is not a member of 'nlohmann::json_abi_v3_11_2::detail::iter_impl<nlohmann::json_abi_v3_11_2::ordered_json>'
```

Any hints on how to solve this?

**Comment by falbrechtskirchinger** [comment, marked as answer]
A few observations upfront:

* Don't use types from the `detail` namespace directly (and `json_abi_v3_11_2` is an inline namespace and should be omitted in this context). I recommend you pass `item.key()` and `item.value()` to `parse_json_item()`.
* You should pass any `nlohmann::basic_json` by (const) reference.
* `lua_pushstring(state, val.value().get<std::string>().c_str())` – I hope `lua_pushstring` creates a copy of the string, otherwise your string will point to invalid memory after the end of the statement. `.get_ref<const std::string&>().c_str()` avoids this issue and skips creating an unnecessary temporary.

For your primary issue, I'll need to take a look at the source and will update this post shortly.

Edit: Can you provide the full error message or at least point out the line that triggers the error? `difference_type` should always be defined.

**Reply by robinsch** [reply]
```
json.hpp(19908,58): error C2039: 'difference_type': is not a member of 'nlohmann::json_abi_v3_11_2::detail::iter_impl<nlohmann::json_abi_v3_11_2::ordered_json>'
FrameScript.cpp(9): message : see declaration of 'nlohmann::json_abi_v3_11_2::detail::iter_impl<nlohmann::json_abi_v3_11_2::ordered_json>'
json.hpp(12843): message : see reference to class template instantiation 'nlohmann::json_abi_v3_11_2::basic_json<nlohmann::json_abi_v3_11_2::ordered_map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::json_abi_v3_11_2::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>' being compiled
json.hpp(5122): message : see reference to class template instantiation 'nlohmann::json_abi_v3_11_2::detail::iter_impl<nlohmann::json_abi_v3_11_2::ordered_json>' being compiled
FrameScript.cpp(12): message : see reference to class template instantiation 'nlohmann::json_abi_v3_11_2::detail::iteration_proxy_value<nlohmann::json_abi_v3_11_2::detail::iter_impl<nlohmann::json_abi_v3_11_2::ordered_json>>' being compiled
```

https://github.com/nlohmann/json/blob/develop/single_include/nlohmann/json.hpp#L19948

FrameScript:9 is the parse_json_item definition.
FrameScript:12 is the first access to passed val

**Reply by robinsch** [reply]
I adjust the code to your suggestions, I am unsure how to pass item.value() (what type should I use to pass it for ordered_json)?

EDIT:

I think I solved the issue by passing key and value, thanks for the fast and good hints!
```
void parse_json_item(std::string const& key, nlohmann::ordered_json::iterator::reference const& value, LuaState* state)
```
Is this the correct way to pass value?

**Reply by falbrechtskirchinger** [reply]
> I adjust the code to your suggestions, I am unsure how to pass item.value() (what type should I use to pass it for ordered_json)?

`item.value()` is the `nlohmann::basic_json` type, i.e., `nlohmann::ordered_json`.

I'm not sure what's causing this error...

**Reply by falbrechtskirchinger** [reply]
> ```
> void parse_json_item(std::string const& key, nlohmann::ordered_json::iterator::reference const& value, LuaState* state)
> ```
>
> Is this the correct way to pass value?

That's fine. You can simplify it as suggested above, but that's not required.

I'm now able to reproduce the core issue. Looks like you uncovered a bug. Glad you were able to work around it!
